src/pythonfinder/models/mixins.py

```diff
@@ -14,47 +21,56 @@
     expand_paths,
     looks_like_python,
     path_is_known_executable,
+    ensure_path,
+    filter_pythons,
+    is_in_path,
+    normalize_path,
 )
 
-if TYPE_CHECKING:
-    from pathlib import Path
-
-    from .path import PathEntry
-    from .python import PythonVersion
-
-    BaseFinderType = TypeVar("BaseFinderType")
-
-
-@attr.s(slots=True)
-class BasePath:
-    path: Path = attr.ib(default=None)
-    _children: dict[str, PathEntry] = attr.ib(default=attr.Factory(dict), order=False)
-    only_python: bool = attr.ib(default=False)
-    name = attr.ib(type=str)
-    _py_version: PythonVersion | None = attr.ib(default=None, order=False)
-    _pythons: DefaultDict[str, PathEntry] = attr.ib(
-        default=attr.Factory(defaultdict), order=False
-    )
-    _is_dir: bool | None = attr.ib(default=None, order=False)
-    _is_executable: bool | None = attr.ib(default=None, order=False)
-    _is_python: bool | None = attr.ib(default=None, order=False)
+from ..environment import get_shim_paths
+
+
+class PathEntry(BaseModel):
+    is_root: bool = Field(default=False, order=False)
+    name: Optional[str] = None
+    path: Optional[Path] = None
+    children_ref: Optional[Any] = Field(default_factory=lambda: dict())
+    only_python: Optional[bool] = False
+    py_version_ref: Optional[Any] = None
+    pythons_ref: Optional[Dict[Any, Any]] = defaultdict(lambda: None)
+    is_dir_ref: Optional[bool] = None
+    is_executable_ref: Optional[bool] = None
+    is_python_ref: Optional[bool] = None
+
+    class Config:
+        validate_assignment = True
+        arbitrary_types_allowed = True
+        allow_mutation = True
+        include_private_attributes = True
+
+    @validator('children', pre=True, always=True, check_fields=False)
+    def set_children(cls, v, values, **kwargs):
+        path = values.get('path')
+        if path:
+            values['name'] = path.name
+        return v or cls()._gen_children()
 
     def __str__(self) -> str:
-        return fs_str(f"{self.path.as_posix()}")
+        return fs_str("{0}".format(self.path.as_posix()))
```

Comment:
I believe `fs_str` can be removed, as well as compat.py. We these all over in vistir and reqlib.
Past issues with py2py3 version and unicode are now now history.